packages/api/tests/lambdas/sf-event-sqs-to-db-records/test-write-execution.js

```diff
@@ -63,119 +57,417 @@ test.beforeEach((t) => {
   } = t.context;
 
-  delete process.env.RDS_DEPLOYMENT_CUMULUS_VERSION;
-  await t.throwsAsync(
-    shouldWriteExecutionToRDS(cumulusMessage, collectionCumulusId, knex)
+  cumulusMessage.meta.status = 'completed';
+
+  const record = buildExecutionRecord({
+    cumulusMessage,
+  });
+
+  t.is(record.status, 'completed');
+  t.is(record.original_payload, undefined);
+  t.deepEqual(record.final_payload, cumulusMessage.payload);
+});
+
+test('buildExecutionRecord returns record with duration', (t) => {
+  const {
+    cumulusMessage,
+    workflowStartTime,
+  } = t.context;
+
+  cumulusMessage.meta.status = 'completed';
+  cumulusMessage.cumulus_meta.workflow_stop_time = workflowStartTime + 5000;
+
+  const record = buildExecutionRecord({
+    cumulusMessage,
+  });
+
+  t.is(record.duration, 5);
+});
+
+test('buildExecutionRecord returns record with error', (t) => {
+  const {
+    cumulusMessage,
+  } = t.context;
+
+  cumulusMessage.meta.status = 'failed';
+  const exception = {
+    Error: new Error('fake error'),
+    Cause: 'an error occurred',
+  };
+  cumulusMessage.exception = exception;
+
+  const record = buildExecutionRecord({
+    cumulusMessage,
+  });
+
+  t.deepEqual(record.error, exception);
+});
+
+test('writeRunningExecutionViaTransaction only updates allowed fields on conflict', async (t) => {
+  const { executionArn, knex } = t.context;
+  const initialDatetime = new Date();
+  const executionRecord = {
+    arn: executionArn,
+    status: 'running',
+    url: 'first-url',
+    original_payload: {
+      key: 'value',
+    },
+    created_at: initialDatetime,
+    updated_at: initialDatetime,
+    timestamp: initialDatetime,
+    workflow_name: 'TestWorkflow',
+  };
+  await knex(tableNames.executions).insert(executionRecord);
+
+  const newPayload = {
+    key2: 'value2',
+  };
+  const newDatetime = new Date();
+  executionRecord.created_at = newDatetime;
+  executionRecord.updated_at = newDatetime;
+  executionRecord.timestamp = newDatetime;
+  executionRecord.original_payload = newPayload;
+  executionRecord.workflow_name = 'TestWorkflow2';
+  executionRecord.url = 'new-url';
+
+  await knex.transaction(
+    (trx) =>
+      writeRunningExecutionViaTransaction({
+        trx,
+        executionRecord,
+      })
+  );
+  const updatedRecord = await knex(tableNames.executions)
+    .where({ arn: executionArn })
+    .first();
+  t.is(updatedRecord.status, 'running');
+  // should have been updated
+  t.deepEqual(updatedRecord.original_payload, newPayload);
+  t.deepEqual(updatedRecord.timestamp, newDatetime);
+  t.deepEqual(updatedRecord.updated_at, newDatetime);
+  t.deepEqual(updatedRecord.created_at, newDatetime);
+  // should not have been updated
+  t.is(updatedRecord.url, 'first-url');
+  t.is(updatedRecord.workflow_name, 'TestWorkflow');
+});
+
+test('writeExecutionViaTransaction() writes record with foreign key references', async (t) => {
+  const { executionArn, cumulusMessage, knex } = t.context;
```

Comment:
Looking at what's being evaluated here.... we're really testing: 

1) That the method under test passes the FKs to buildExecutionRecord (the actual building of the record should be tested in that methods tests)... which it should do each time. 
2) That knex can insert whatever comes back from #1 and returns the expected ID, and in this instance in particular it includes *a specific FK* 

Is there a reason we shouldn't make all of the FK relationships a setup for each write test below and just remove this test entirely (possibly more rad) , or alternately mock out the call to build record and test that it's called with the expected values (perhaps less rad)?      I'm not sure there's a valid scenario these tests cover where the reverse case (no FKs) fails the knex insert where a FK constrained object *would*. 

🤔

Reply:
I settled on this: 9e627ab56810417dc113c9394c0cd0128c2e4f78